src/main/java/org/jenkinsci/plugins/workflow/cps/Snippetizer.java

```diff
@@ -62,105 +65,235 @@
  */
 @Extension public class Snippetizer implements RootAction, DescriptorByNameOwner {
-    
+
+    /**
+     * Short-hand for the top-level invocation.
+     */
     static String object2Groovy(Object o) throws UnsupportedOperationException {
-        Class<? extends Object> clazz = o.getClass();
+        return object2Groovy(new StringBuilder(),o, false).toString();
+    }
+
+    /**
+     * Renders the invocation syntax to re-create a given object 'o' into 'b'
+     *
+     * @param nestedExp
+     *      true if this object is written as a nested expression (in which case we always produce parenthesis for readability
+     * @return  the same object as 'b'
+     */
+    static StringBuilder object2Groovy(StringBuilder b, Object o, boolean nestedExp) throws UnsupportedOperationException {
+        if (o == null) {
+            return b.append("null");
+        }
+        final Class<?> clazz = o.getClass();
+
+        if (clazz == String.class || clazz == Character.class) {
+            String text = String.valueOf(o);
+            if (text.contains("\n")) {
+                b.append("'''").append(text.replace("\\", "\\\\").replace("'", "\\'")).append("'''");
+            } else {
+                b.append('\'').append(text.replace("\\", "\\\\").replace("'", "\\'")).append('\'');
+            }
+            return b;
+        }
+
+        if (clazz == Boolean.class || clazz == Integer.class || clazz == Long.class) {
+            return b.append(o);
+        }
+
+        if (o instanceof List) {
+            return list2groovy(b, (List<?>) o);
+        }
+
+        if (o instanceof Map) {
+            return map2groovy(b, (Map) o);
+        }
+
+        if (o instanceof UninstantiatedDescribable) {
+            return ud2groovy(b,(UninstantiatedDescribable)o, false, nestedExp);
+        }
+
         for (StepDescriptor d : StepDescriptor.all()) {
             if (d.clazz.equals(clazz)) {
-                StringBuilder b = new StringBuilder(d.getFunctionName());
                 Step step = (Step) o;
-                Map<String,Object> args = new TreeMap<String,Object>(d.defineArguments(step));
-                boolean first = true;
-                boolean singleMap = args.size() == 1 && args.values().iterator().next() instanceof Map;
-                for (Map.Entry<String,Object> entry : args.entrySet()) {
-                    if (first) {
-                        first = false;
-                        if (d.takesImplicitBlockArgument() || singleMap) {
-                            b.append('(');
-                        } else {
-                            b.append(' ');
+                UninstantiatedDescribable uninst = d.uninstantiate(step);
+
+                if (d.isMetaStep()) {
+                    // if we cannot represent this 'o' in a concise syntax that hides meta-step, set this to true
+                    boolean failSimplification = false;
```

Comment:
I think there are some cases you are forgetting here. For example, there is a `@Symbol` but it clashes with a `StepDescriptor.functionName`. Or there are two different metasteps which can both accept the same `@Symbol`.